### PR TITLE
Added utility for dumping Spring Boot config metadata properties

### PR DESCRIPTION
**Pull request by itzg**
# What

This was my hack day project to give us a way to interrogate a Spring Boot app for all its known configuration properties including the environment variable name for each.

I manually tested this by adding the code shown in the javadoc to the ambassador's main method and invoked it passing `--dump-config=ambassador`, which produced the output:

```
Name | Type | Default | Description | Environment Variable
-----|------|---------|-------------|---------------------
ambassador.alt-external-names | List of String | 127.0.0.1 |  | AMBASSADOR_ALTEXTERNALNAMES
ambassador.cert-chain-path | String | certs/out/ambassador.pem |  | AMBASSADOR_CERTCHAINPATH
ambassador.disable-tls | Boolean | false | FOR DEVELOPMENT ONLY, disable TLS including mutual TLS authentication | AMBASSADOR_DISABLETLS
ambassador.envoy-lease-sec | Long | 30 |  | AMBASSADOR_ENVOYLEASESEC
ambassador.envoy-refresh-interval | String | PT10S |  | AMBASSADOR_ENVOYREFRESHINTERVAL
ambassador.envoy-refresh-parallelism | Long | 5 |  | AMBASSADOR_ENVOYREFRESHPARALLELISM
ambassador.external-name | String | localhost |  | AMBASSADOR_EXTERNALNAME
ambassador.key-path | String | certs/out/ambassador-pkcs8-key.pem |  | AMBASSADOR_KEYPATH
ambassador.max-api-retry-attempts | Integer | 10 | When calling APIs of other microservices, this will be the max attempts retried | AMBASSADOR_MAXAPIRETRYATTEMPTS
ambassador.public-zone-tenants | List of String |  | Only Envoys of these tenants are allowed to advertise a zone with prefixed with 'publicZonePrefix' | AMBASSADOR_PUBLICZONETENANTS
ambassador.trust-cert-path | String | certs/out/ca.pem |  | AMBASSADOR_TRUSTCERTPATH
ambassador.vault-pki-role | String | telemetry-infra |  | AMBASSADOR_VAULTPKIROLE
```

which actually renders in markdown as

Name | Type | Default | Description | Environment Variable
-----|------|---------|-------------|---------------------
ambassador.alt-external-names | List of String | 127.0.0.1 |  | AMBASSADOR_ALTEXTERNALNAMES
ambassador.cert-chain-path | String | certs/out/ambassador.pem |  | AMBASSADOR_CERTCHAINPATH
ambassador.disable-tls | Boolean | false | FOR DEVELOPMENT ONLY, disable TLS including mutual TLS authentication | AMBASSADOR_DISABLETLS
ambassador.envoy-lease-sec | Long | 30 |  | AMBASSADOR_ENVOYLEASESEC
ambassador.envoy-refresh-interval | String | PT10S |  | AMBASSADOR_ENVOYREFRESHINTERVAL
ambassador.envoy-refresh-parallelism | Long | 5 |  | AMBASSADOR_ENVOYREFRESHPARALLELISM
ambassador.external-name | String | localhost |  | AMBASSADOR_EXTERNALNAME
ambassador.key-path | String | certs/out/ambassador-pkcs8-key.pem |  | AMBASSADOR_KEYPATH
ambassador.max-api-retry-attempts | Integer | 10 | When calling APIs of other microservices, this will be the max attempts retried | AMBASSADOR_MAXAPIRETRYATTEMPTS
ambassador.public-zone-tenants | List of String |  | Only Envoys of these tenants are allowed to advertise a zone with prefixed with 'publicZonePrefix' | AMBASSADOR_PUBLICZONETENANTS
ambassador.trust-cert-path | String | certs/out/ca.pem |  | AMBASSADOR_TRUSTCERTPATH
ambassador.vault-pki-role | String | telemetry-infra |  | AMBASSADOR_VAULTPKIROLE

### CODE REVIEW

File: src/main/java/com/rackspace/salus/common/util/DumpConfigProperties.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+          dumpConfig(Arrays.asList(patterns));
+        }
+      }
+    }
+
+    return new Chained();
+  }
+
+  public static class Chained {
+    public void or(Runnable runnable) {
```
Comment (GeorgeJahad) [inline src/main/java/com/rackspace/salus/common/util/DumpConfigProperties.java]: i don't quite get the point of the "or" method.

Comment (itzg) [reply src/main/java/com/rackspace/salus/common/util/DumpConfigProperties.java]: Yes, it's very weird to see it in isolation and even looking at it in the context of calling it I'm not totally sure I like it:

```
@SpringBootApplication
@EnableScheduling
@EnableEtcd
@EnableSalusKafkaMessaging
public class TelemetryAmbassadorApplication {

    public static void main(String[] args) {
        DumpConfigProperties.process(args)
            .or(() -> {
                SpringApplication.run(TelemetryAmbassadorApplication.class, args);
            });
    }
}
```

Comment (itzg) [reply src/main/java/com/rackspace/salus/common/util/DumpConfigProperties.java]: ...originally I was thinking of calling is `then(...)` but I realized with the `System.exit()` call in there `or(...)` would better clarify the early exit.

Comment (itzg) [reply src/main/java/com/rackspace/salus/common/util/DumpConfigProperties.java]: ...and what I really wanted to use in the first place was https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-spring-application.html#boot-features-command-line-runner but I discovered those called very, very late after all the application context is started.

Comment (GeorgeJahad) [reply src/main/java/com/rackspace/salus/common/util/DumpConfigProperties.java]: ok, that makes sense, thanks

**Comment by jjbuchan** [comment]
I've not looked at the code but this is very cool!

File: src/main/java/com/rackspace/salus/common/util/DumpConfigProperties.java (hunk 1 above)
Comment (itzg) [reply src/main/java/com/rackspace/salus/common/util/DumpConfigProperties.java]: I ended up changing my mind and removing the chaining-or stuff. When looking at a stack trace while debugging I didn't like seeing DumpConfigProperties sitting there at the bottom of the stack.

**Pull request merged**